File: macOS/Packages/ClaudeUsageCore/Sources/ClaudeUsageCore/PricingTable.swift
import Foundation

public struct ModelPricing: Sendable {
    public let inputPerMillion: Double
    public let outputPerMillion: Double
    public let cacheWritePerMillion: Double
    public let cacheReadPerMillion: Double

    public init(inputPerMillion: Double, outputPerMillion: Double, cacheWritePerMillion: Double, cacheReadPerMillion: Double) {
        self.inputPerMillion = inputPerMillion
        self.outputPerMillion = outputPerMillion
        self.cacheWritePerMillion = cacheWritePerMillion
        self.cacheReadPerMillion = cacheReadPerMillion
    }
}

/// Approximate USD-per-1M-token pricing, used only for the "estimated cost" display.
///
/// NOTE: these numbers can go stale — Anthropic can change pricing at any time, and
/// this table was filled in from the best information available while building this
/// app, not fetched live. Verify against https://anthropic.com/pricing and update the
/// entries below if they drift. Models not listed here are simply excluded from cost
/// totals (never silently priced at $0).
public enum PricingTable {
    public static let byModel: [String: ModelPricing] = [
        "claude-opus-5": ModelPricing(inputPerMillion: 15, outputPerMillion: 75, cacheWritePerMillion: 18.75, cacheReadPerMillion: 1.5),
        "claude-sonnet-5": ModelPricing(inputPerMillion: 3, outputPerMillion: 15, cacheWritePerMillion: 3.75, cacheReadPerMillion: 0.3),
        "claude-haiku-4-5-20251001": ModelPricing(inputPerMillion: 1, outputPerMillion: 5, cacheWritePerMillion: 1.25, cacheReadPerMillion: 0.1),
    ]

    public static func estimatedCostUSD(
        model: String,
        inputTokens: Int,
        outputTokens: Int,
        cacheCreationTokens: Int,
        cacheReadTokens: Int
    ) -> Double? {
        guard let pricing = byModel[model] else { return nil }
        return Double(inputTokens) / 1_000_000 * pricing.inputPerMillion
            + Double(outputTokens) / 1_000_000 * pricing.outputPerMillion
            + Double(cacheCreationTokens) / 1_000_000 * pricing.cacheWritePerMillion
            + Double(cacheReadTokens) / 1_000_000 * pricing.cacheReadPerMillion
    }
}

File: TokenMihariban/Models/PricingTable.cs
using System.Collections.Generic;

namespace TokenMihariban.Models;

public sealed record ModelPricing(double InputPerMillion, double OutputPerMillion, double CacheWritePerMillion, double CacheReadPerMillion);

/// <summary>
/// Approximate USD-per-1M-token pricing, used only for the "estimated cost" display.
/// NOTE: these numbers can go stale — verify against https://anthropic.com/pricing.
/// Models not listed here are simply excluded from cost totals (never silently priced at $0).
/// </summary>
public static class PricingTable
{
    public static readonly Dictionary<string, ModelPricing> ByModel = new()
    {
        ["claude-opus-5"] = new ModelPricing(15, 75, 18.75, 1.5),
        ["claude-sonnet-5"] = new ModelPricing(3, 15, 3.75, 0.3),
        ["claude-haiku-4-5-20251001"] = new ModelPricing(1, 5, 1.25, 0.1),
    };

    public static double? EstimatedCostUSD(string model, long inputTokens, long outputTokens, long cacheCreationTokens, long cacheReadTokens)
    {
        if (!ByModel.TryGetValue(model, out var pricing)) return null;
        return inputTokens / 1_000_000.0 * pricing.InputPerMillion
            + outputTokens / 1_000_000.0 * pricing.OutputPerMillion
            + cacheCreationTokens / 1_000_000.0 * pricing.CacheWritePerMillion
            + cacheReadTokens / 1_000_000.0 * pricing.CacheReadPerMillion;
    }
}
